README.md
![Hoa](http://static.hoa-project.net/Image/Hoa_small.png)

[![MELPA](http://melpa.org/packages/hoa-pp-mode-badge.svg)](http://melpa.org/#/hoa-pp-mode)
[![MELPA Stable](http://stable.melpa.org/packages/hoa-pp-mode-badge.svg)](http://stable.melpa.org/#/hoa-pp-mode)


Hoa is a **modular**, **extensible** and **structured** set of PHP libraries.
Moreover, Hoa aims at being a bridge between industrial and research worlds.

# Contributions/Emacs/Pp

This repository contains tools for the PP grammar description language from
[`Hoa\Compiler`](http://central.hoa-project.net/Resource/Library/Compiler).

It provides a major mode for editing PP grammars with the following features:

- syntax coloration
- auto-indentation
- Imenu support

![PP Major mode screenshot](http://central.hoa-project.net/Resource/Contributions/Emacs/Pp/screenshots/sample.png?format=raw)

## Installation

If you have setup MELPA or MELPA stable repositories, type `M-x package-install
hoa-pp-mode`.

If you want to install it by hand, you must first install
[names](https://github.com/Malabarba/names).

Then, clone this repository on your computer.

```sh
git clone git@github.com:hoaproject/Contributions-Emacs-Pp.git hoa-pp-mode
```

Finally, add this code in your Emacs configuration:

```emacs-lisp
(add-to-list 'load-path "/path/to/hoa-pp-mode")
(require 'hoa-pp-mode)
```

## Quick usage

Simply open a `.pp` file with Emacs. `hoa-pp-mode` will automatically be enabled.

## Documentation

Different documentations can be found on the website:
[http://hoa-project.net/](http://hoa-project.net/).

## License

Hoa is under the New BSD License (BSD-3-Clause). Please, see
[`LICENSE`](http://hoa-project.net/LICENSE).


CHANGELOG.md
# Upcoming

* Add support for Imenu (Steven Rémot, 2015-10-21T18:45:17 2015+02:00)

# 0.2.1

* Adapt README and code to the new repsoitory (Ivan Enderlin, Julien Bianchi, Steven Rémot)
* Remove calls to setq-local (Steven Rémot, 2015-08-18T21:44:23+02:00)

# 0.2.0

* Add Cask file (Steven Rémot, 2015-08-14T13:45:13+02:00)
* Implement auto-indentation (Steven Rémot, 2015-08-14T13:36:28+02:00)
* Add a screenshot to the README (Steven Rémot, 2015-08-11T13:20:37+02:00)
* Add missing import to names (Steven Rémot, 2015-08-11T13:18:36+02:00)
* Improve description (Steven Rémot, 2015-08-06T13:59:21+02:00)
* Add URL in source meta-data (Steven Rémot, 2015-08-06T15:58:15T02:00)

# 0.1

First snapshot, syntaxical coloration.


hoa-pp-mode.el
;;; hoa-pp-mode.el --- Major mode for Hoa PP grammars -*- lexical-binding: t -*-

;; Copyright © 2015-2015, Hoa community. All rights reserved.

;; Author: Steven Rémot
;; Maintainer: Steven Rémot
;; Version: 0.2.1
;; Keywords: php, hoa
;; URL: https://github.com/hoaproject/Contributions-Emacs-Pp
;; Package-Requires: ((emacs "24.1") (names "20150723.0"))

;; This file is not part of GNU Emacs.

;; Redistribution and use in source and binary forms, with or without
;; modification, are permitted provided that the following conditions are met:
;;     * Redistributions of source code must retain the above copyright
;;       notice, this list of conditions and the following disclaimer.
;;     * Redistributions in binary form must reproduce the above copyright
;;       notice, this list of conditions and the following disclaimer in the
;;       documentation and/or other materials provided with the distribution.
;;     * Neither the name of the Hoa nor the names of its contributors may be
;;       used to endorse or promote products derived from this software without
;;       specific prior written permission.

;; THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
;; AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
;; IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;; ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
;; LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
;; CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
;; SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
;; CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
;; ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
;; POSSIBILITY OF SUCH DAMAGE.

;;; Commentary:
;;
;; This package provides a major mode for editing hoa *.pp files.
;; Its current features are:
;; - syntax coloration
;; - auto-indentation
;; - Imenu support

;;; Code:
(eval-when-compile (require 'names))

(define-namespace hoa-pp-mode-

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Syntax ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defconst directive-regexp (rx bol "%" (group (or "token" "skip"))
                               (? (+ blank) (group (+ (not (any blank ?\n))))
                                  (? (+ blank) (group  (+ (or (not (any ?\n ?-))
                                                              (seq ?- (not (any ?> ?\n)))))))))
  "Regular expression for matching compiler directives.")

(defconst arrow-directive-regexp (rx bol "%token"
                                     (+ (or (not (any ?\n ?-))
                                            (seq ?- (not (any ?>)))))
                                     (group "->")
                                     (? (* blank) (group (+ (not (any blank ?\n)))))
                                     eol)
  "Regular expression for matching arrow clause behind comiler directives.")

(defconst rule-regexp (rx bol (? "#") (group (+ (or word (syntax symbol)))) ":" (* space) eol)
  "Regular expression for matching rule declaration.")

(defconst font-lock-keywords
  `(;; Compiler directives
    (,hoa-pp-mode-directive-regexp . (1 font-lock-keyword-face))
    (,hoa-pp-mode-directive-regexp . (2 font-lock-constant-face))
    (,hoa-pp-mode-directive-regexp . (3 font-lock-string-face))
    (,hoa-pp-mode-arrow-directive-regexp . (1 font-lock-builtin-face))
    (,hoa-pp-mode-arrow-directive-regexp . (2 font-lock-constant-face))

    ;; Rules
    (,(rx bol (* space) (? "#") (group (+ (or word (syntax symbol)))) ":" (* space) eol)  . (1 font-lock-function-name-face))

    ;; Token and rule use
    (, (rx "::" (group (+ (or word (syntax symbol)))) "::") . (1 font-lock-constant-face))
    (, (rx "<" (group (+ (or word (syntax symbol)))) ">") . (1 font-lock-constant-face))
    (, (rx bow (group (+ (or word (syntax symbol)))) "()") . (1 font-lock-function-name-face))
    )
  "Keywords highlighting for Hoa PP mode.")

(defun setup-syntax-table (&optional table)
  "Setup the syntax table for `hoa-pp-mode'.

TABLE is a syntax table.  It will be the default table if not provided."
  (modify-syntax-entry ?\" "." table)
  (modify-syntax-entry ?/ "<12" table)
  (modify-syntax-entry ?\n ">" table))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Indentation ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defun is-at-directive-start? ()
  "Return t if the current point is at a directive start."
  (= ?% (aref (buffer-substring-no-properties (point) (1+ (point))) 0)))

(defun is-at-rule-start? ()
  "Return t if the current point is at a rule start."
  (looking-at rule-regexp))

(defun is-in-rule? ()
  "Return t if the current point is inside a rule definition."
  (catch 'in-rule?
    (save-excursion
      (forward-line -1)
      (beginning-of-line)
      (while (not (= (point) (point-min)))
        (back-to-indentation)
        (when (is-at-directive-start?)
          (throw 'in-rule? nil))
        (when (is-at-rule-start?)
          (throw 'in-rule? t))
        (forward-line -1)
        (beginning-of-line)))))

(defun ensure-in-text ()
  "Go back to indentation of point is before line's text."
  (let ((text-start (save-excursion
                      (back-to-indentation)
                      (point))))
    (when (< (point) text-start)
      (back-to-indentation))))

(defun indent-line ()
  "Indent the current line."
  (interactive)
  (save-excursion
    (back-to-indentation)
    (cond
     ((is-at-directive-start?)
      (beginning-of-line)
      (just-one-space 0))
     ((is-at-rule-start?)
      (beginning-of-line)
      (just-one-space 0))
     ((is-in-rule?)
      (indent-to (indent-next-tab-stop 0)))
     (t
      (indent-to 0))))
  (ensure-in-text))

(defun setup-indentation ()
  "Setup the indent function for `hoa-pp-mode'."
  (set (make-local-variable 'indent-line-function) #'indent-line)
  (set (make-local-variable 'electric-indent-chars) '(?\n ?:)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Imenu ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defconst imenu-expressions (list
                             (list "Tokens" (rx bol "%token" (+ blank) (group (+ (not (any blank ?\n))))) 1)
                             (list "Rules" rule-regexp 1))
  "Imenu configuration for `hoa-pp-mode'.")

(defun setup-imenu ()
  "Configure Imenu for `hoa-pp-mode'.

Provide menu section for tokens and rules."
  (setq imenu-generic-expression imenu-expressions)
  (imenu-add-menubar-index))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Mode definition ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


;;;###autoload
(define-derived-mode hoa-pp-mode
  prog-mode "Hoa PP"
  "Major mode for editing Hoa PP grammars.
\\{hoa-pp-mode-map"
  (set (make-local-variable 'font-lock-defaults) '(hoa-pp-mode-font-lock-keywords))
  (hoa-pp-mode-setup-syntax-table)
  (hoa-pp-mode-setup-indentation)
  (hoa-pp-mode-setup-imenu))

;;;###autoload
(add-to-list 'auto-mode-alist '("\\.pp$" . hoa-pp-mode))

(provide 'hoa-pp-mode)

;;; hoa-pp-mode.el ends here


pp.el
;;; pp.el --- Major mode for Hoa PP grammars -*- lexical-binding: t -*-
;;
;; Copyright (C) 2015 Steven Rémot
;;
;; Author: Steven Rémot
;; Version: 0.1
;; Keywords: php, hoa
;; URL: https://github.com/hoaproject/Contributions-Emacs-Pp/
;; Package-Requires: ((emacs "24.1") (names "20150723.0"))
;;
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
;;
;;; Commentary:
;;
;; This package provides a major mode for editing hoa *.pp files.
;; It currently has syntax coloration and auto-indentation.

;;; Code:
(eval-when-compile (require 'names))

(define-namespace hoa-pp-mode-

(defconst directive-regexp (rx bol "%" (group (or "token" "skip"))
                               (? (+ blank) (group (+ (not (any blank ?\n))))
                                  (? (+ blank) (group  (+ (or (not (any ?\n ?-))
                                                              (seq ?- (not (any ?> ?\n)))))))))
  "Regular expression for matching compiler directives.")

(defconst arrow-directive-regexp (rx bol "%token"
                                     (+ (or (not (any ?\n ?-))
                                            (seq ?- (not (any ?>)))))
                                     (group "->")
                                     (? (* blank) (group (+ (not (any blank ?\n)))))
                                     eol)
  "Regular expression for matching arrow clause behind comiler directives.")

(defconst rule-regexp (rx (? "#") (group (+ (or word (syntax symbol)))) ":" (* space) eol)
  "Regular expression for matching rule declaration.")

(defconst font-lock-keywords
  `(;; Compiler directives
    (,hoa-pp-mode-directive-regexp . (1 font-lock-keyword-face))
    (,hoa-pp-mode-directive-regexp . (2 font-lock-constant-face))
    (,hoa-pp-mode-directive-regexp . (3 font-lock-string-face))
    (,hoa-pp-mode-arrow-directive-regexp . (1 font-lock-builtin-face))
    (,hoa-pp-mode-arrow-directive-regexp . (2 font-lock-constant-face))

    ;; Rules
    (,(rx bol (* space) (? "#") (group (+ (or word (syntax symbol)))) ":" (* space) eol)  . (1 font-lock-function-name-face))

    ;; Token and rule use
    (, (rx "::" (group (+ (or word (syntax symbol)))) "::") . (1 font-lock-constant-face))
    (, (rx "<" (group (+ (or word (syntax symbol)))) ">") . (1 font-lock-constant-face))
    (, (rx bow (group (+ (or word (syntax symbol)))) "()") . (1 font-lock-function-name-face))
    )
  "Keywords highlighting for Hoa PP mode.")

(defun setup-syntax-table (&optional table)
  "Setup the syntax table for `hoa-pp-mode'.
TABLE is a syntax table.  It will be the default table if not provided."
  (modify-syntax-entry ?\" "." table)
  (modify-syntax-entry ?/ "<12" table)
  (modify-syntax-entry ?\n ">" table))

(defun is-at-directive-start? ()
  "Return t if the current point is at a directive start."
  (= ?% (aref (buffer-substring-no-properties (point) (1+ (point))) 0)))

(defun is-at-rule-start? ()
  "Return t if the current point is at a rule start."
  (looking-at rule-regexp))

(defun is-in-rule? ()
  "Return t if the current point is inside a rule definition."
  (catch 'in-rule?
    (save-excursion
      (forward-line -1)
      (beginning-of-line)
      (while (not (= (point) (point-min)))
        (back-to-indentation)
        (when (is-at-directive-start?)
          (throw 'in-rule? nil))
        (when (is-at-rule-start?)
          (throw 'in-rule? t))
        (forward-line -1)
        (beginning-of-line)))))

(defun ensure-in-text ()
  "Go back to indentation of point is before line's text."
  (let ((text-start (save-excursion
                      (back-to-indentation)
                      (point))))
    (when (< (point) text-start)
      (back-to-indentation))))

(defun indent-line ()
  "Indent the current line."
  (interactive)
  (save-excursion
    (back-to-indentation)
    (cond
     ((is-at-directive-start?)
      (beginning-of-line)
      (just-one-space 0))
     ((is-at-rule-start?)
      (beginning-of-line)
      (just-one-space 0))
     ((is-in-rule?)
      (indent-to (indent-next-tab-stop 0)))
     (t
      (indent-to 0))))
  (ensure-in-text))

(defun setup-indentation ()
  "Setup the indent function for `hoa-pp-mode'."
  (set (make-local-variable 'indent-line-function) #'indent-line)
  (set (make-local-variable 'electric-indent-chars) '(?\n ?:)))
)

;;;###autoload
(define-derived-mode hoa-pp-mode
  prog-mode "Hoa PP"
  "Major mode for editing Hoa PP grammars.
\\{hoa-pp-mode-map"
  (set (make-local-variable 'font-lock-defaults) '(hoa-pp-mode-font-lock-keywords))
  (hoa-pp-mode-setup-syntax-table)
  (hoa-pp-mode-setup-indentation))

;;;###autoload
(add-to-list 'auto-mode-alist '("\\.pp$" . hoa-pp-mode))

(provide 'hoa-pp-mode)

;;; pp.el ends here
